### PR TITLE
Prep for extra plugins

### PR DESCRIPTION
**Pull request by coltonmorris**
Created an `Outputter` interface that must be fullfilled for each output type (redux3, redux4, etc.. ). Already implemented a struct that fullfills it generically, so if an output was made that uses the same entities then it can use the generic one, otherwise it can just create it's own (shown by the mobx example).
More details can be found by reading the `CONTRIBUTING.md` file.

Implements #3 and #5

### CODE REVIEW

**Comment by coltonmorris** [comment]
Should note that this has breaking changes because of #5.

now file options are defined like this (note the new top level key "state_file_options"):
```

option (state_file_options) = {
  debounce: 510,
  debug: false,
  default_retries: 0,
  default_timeout: 3000,
  hostname_location: "config.host",
  port: 9090,
  protoc_ts_path: "../",
  output_type: redux3
  /* option (auth_token_location) = "config.token"; */
};
```
and field options are defined like this (note the new top level key "state_field_options"):
```
  repeated readinglist.Book library = 1 [
    (state_field_options).method.create = "readinglist.ReadingList.CreateBook",
    (state_field_options).method.update = "readinglist.ReadingList.UpdateBook",
    (state_field_options).method.delete = "readinglist.ReadingList.DeleteBook",
    (state_field_options).method.list = "readinglist.ReadingList.ReadAllBooks"
  ];
```

File: e2e/mobx/protos/basic.proto [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+option (state_file_options) = {
+  debounce: 510,
+  debug: false,
+  default_retries: 0,
+  default_timeout: 3000,
+  hostname_location: "config.host",
+  port: 9090,
+  protoc_ts_path: "../",
+  output_type: mobx
+  /* option (auth_token_location) = "config.token"; */
```
Comment (hectim) [inline e2e/mobx/protos/basic.proto]: `auth_token_location = "config.token"`

File: e2e/redux3/protos/basic.proto [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+option (state_file_options) = {
+  debounce: 510,
+  debug: false,
+  default_retries: 0,
+  default_timeout: 3000,
+  hostname_location: "config.host",
+  port: 9090,
+  protoc_ts_path: "../",
+  output_type: redux3
+  /* option (auth_token_location) = "config.token"; */
```
Comment (hectim) [inline e2e/redux3/protos/basic.proto]: `auth_token_location = "config.token"`

File: e2e/redux3/spec/helpers/alias.js [hunk 3]
```diff
@@ -1,8 +1,8 @@
 const moduleAlias = require('module-alias');
 
 moduleAlias.addAliases({
-  'proto': '../protos/',
-  '@App': 'neo/'
+  'protos': '../protos/',
+  '@App': '../'
```
Comment (hectim) [inline e2e/redux3/spec/helpers/alias.js]: Ok I wasn't sure what you meant about the tests still being from neo. 👌

File: e2e/redux3/src/public/index.html [hunk 4]
```diff
@@ -12,7 +12,7 @@
     You need to enable JavaScript to run this app.
   </noscript>
   <div id="root"></div>
-  <script type="text/javascript" src="bundle.js" charset="utf-8"></script>
+  <!-- <script type="text/javascript" src="bundle.js" charset="utf-8"></script> -->
```
Comment (hectim) [inline e2e/redux3/src/public/index.html]: The purpose of this? Just a placeholder page?

File: e2e/redux4/protos/basic.proto [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+option (state_file_options) = {
+  debounce: 510,
+  debug: false,
+  default_retries: 0,
+  default_timeout: 3000,
+  hostname_location: "config.host",
+  port: 9090,
+  protoc_ts_path: "../",
+  output_type: redux4
+  /* option (auth_token_location) = "config.token"; */
```
Comment (hectim) [inline e2e/redux4/protos/basic.proto]: `auth_token_location = "config.token"`

File: generator/action.go [hunk 6]
```diff
@@ -333,16 +164,18 @@ func CreateActionFile(stateFields []*gp.FieldDescriptorProto, customFields []*gp
...
-	updateT := template.Must(template.New("update").Funcs(funcMap).Parse(updateTemplate))
-	customT := template.Must(template.New("update").Funcs(funcMap).Parse(customTemplate))
+  importsT := template.Must(template.New("imports").Funcs(funcMap).Parse(this.ActionFile.ImportTemplate))
+	getT := template.Must(template.New("get").Funcs(funcMap).Parse(this.ActionFile.GetTemplate))
+	listT := template.Must(template.New("list").Funcs(funcMap).Parse(this.ActionFile.ListTemplate))
+	resetT := template.Must(template.New("reset").Funcs(funcMap).Parse(this.ActionFile.ResetTemplate))
+	createT := template.Must(template.New("create").Funcs(funcMap).Parse(this.ActionFile.CreateTemplate))
+	deleteT := template.Must(template.New("delete").Funcs(funcMap).Parse(this.ActionFile.DeleteTemplate))
+	updateT := template.Must(template.New("update").Funcs(funcMap).Parse(this.ActionFile.UpdateTemplate))
+	customT := template.Must(template.New("update").Funcs(funcMap).Parse(this.ActionFile.CustomTemplate))
```
Comment (hectim) [inline generator/action.go]: `template.New("custom")`
Not your mistake, but mine. The name is pretty arbitrary, but should probably match the variable name.

File: e2e/redux3/spec/helpers/alias.js (hunk 3 above)
Comment (coltonmorris) [reply e2e/redux3/spec/helpers/alias.js]: There was the left over file as well in the spec dir that confused me. Turns out it wasn't being used.

File: generator/aggregator.go [hunk 7]
```diff
@@ -99,7 +78,7 @@ type ServiceExport struct {
 	Exports []string
 }
 
-func CreateAggregateServicesFile(serviceFiles []*gp.FileDescriptorProto, protocTsPath string, statePkg string) (*File, error) {
+func (this *GenericOutputter) CreateAggregateServicesFile(serviceFiles []*gp.FileDescriptorProto, outputType state.OutputTypes, protocTsPath string, statePkg string) (*File, error) {
```
Comment (hectim) [inline generator/aggregator.go]: What is `outputType` used for? It doesn't look like the parameter is used.

Comment (coltonmorris) [reply generator/aggregator.go]: ah, an artifact from my earlier investigations of how to get this architecture to work. Nice catch

File: generator/annotations.go [hunk 8]
```diff
@@ -36,143 +36,40 @@ import (
...
+  }
+
+  if proto.HasExtension(desc.GetOptions(), extName) {
+    pkg, err := proto.GetExtension(desc.GetOptions(), extName)
+    if err != nil {
+      return nil, err
+    }
+    return pkg.(*state.StateFieldOptions), nil
+  }
+  return nil, nil
```
Comment (hectim) [inline generator/annotations.go]: giphy

Bravo

Comment (coltonmorris) [reply generator/annotations.go]: 🎉

File: generator/epic.go [hunk 9]
```diff
@@ -189,23 +80,17 @@ func CreateEpicFile(stateFields []*gp.FieldDescriptorProto, customFields []*gp.F
...
 			return nil, fmt.Errorf("Error getting field level annotations: %v", err)
 		}
 
 		// field level overrides for timeout/retry
-		timeout, err := GetFieldAnnotationInt(field, state.E_Timeout)
-		if err != nil {
-			return nil, fmt.Errorf("Error getting field level timeout annotation: %v", err)
-		}
+    timeout := fieldAnnotations.GetTimeout()
 		if timeout == -1 { // if it wasn't overriden
```
Comment (hectim) [inline generator/epic.go]: I'm not sure these integer annotations return -1 if not set anymore. That was the value assigned in GetFieldOptions, but those methods are more generic now.

File: generator/epic.go [hunk 10]
```diff
@@ -277,30 +161,23 @@ func CreateEpicFile(stateFields []*gp.FieldDescriptorProto, customFields []*gp.F
...
 			return nil, fmt.Errorf("Error getting field level annotations: %v", err)
 		}
 
-		// field level overrides for timeout/retry
-		timeout, err := GetFieldAnnotationInt(field, state.E_Timeout)
-		if err != nil {
-			return nil, fmt.Errorf("Error getting field level timeout annotation: %v", err)
-		}
+		timeout := fieldAnnotations.GetTimeout()
 		if timeout == -1 { // if it wasn't overriden
```
Comment (hectim) [inline generator/epic.go]: missed one

**Comment by hectim** [comment]
The tests for redux3 aren't passing for me.  First problem was ts-node not being there in the node_modules so I added a `yarn` to the test command:

``` "test": "yarn && yarn run test-js && yarn run test-ts",```

and changed all the `import x from y` commands to `requires` and now I get a bunch of type errors:

```src/action.test.ts(105,41): error TS2694: Namespace '"/protoc-gen-state/e2e/redux3/protos/BasicState/readinglist_aggregate"' has no exported member 'Book'.```

I can push my changes and talk with you tomorrow.

**Comment by coltonmorris** [comment]
Now fulfilling #1. Figured I'd pull that into this since they are pretty related.

**Pull request merged**